jacodb-ets/src/main/kotlin/org/jacodb/ets/utils/BlockCfgBuilder.kt

```diff
@@ -0,0 +1,292 @@
+import org.jacodb.ets.model.EtsUnknownType
+
+fun BlockCfg.toEtsBlockCfg(method: EtsMethod): EtsBlockCfg {
+    return EtsBlockCfgBuilder(method).build(this)
+}
+
+class EtsBlockCfgBuilder(
+    val method: EtsMethod,
+) {
+    fun build(blockCfg: BlockCfg): EtsBlockCfg {
+        return EtsBlockCfg(
+            blocks = blockCfg.blocks.map { it.toEtsBasicBlock() },
+            successors = blockCfg.successors,
+        )
+    }
+
+    private var freeTempLocal: Int = 0
+    private fun newTempLocal(): EtsLocal {
+        return EtsLocal(
+            name = "_tmp${freeTempLocal++}",
+        )
+    }
+
+    private fun Block.toEtsBasicBlock(): BasicBlock {
+        val etsStatements: MutableList<EtsStmt> = mutableListOf()
+
+        fun ensureLocal(entity: EtsEntity): EtsLocal {
+            if (entity is EtsLocal) {
+                return entity
+            }
+            val newLocal = newTempLocal()
+            etsStatements += EtsAssignStmt(
+                location = stub,
+                lhv = newLocal,
+                rhv = entity,
+            )
+            return newLocal
+        }
+
+        for (stmt in statements) {
+            when (stmt) {
+                BlockNop -> {
+                    etsStatements += EtsNopStmt(location = stub)
+                }
+
+                is BlockAssign -> {
+                    val lhv = stmt.target.toEtsEntity() as EtsLocal // safe cast
+                    val rhv = stmt.expr.toEtsEntity()
+                    etsStatements += EtsAssignStmt(
+                        location = stub,
+                        lhv = lhv,
+                        rhv = rhv,
+                    )
+                }
+
+                is BlockReturn -> {
+                    val returnValue = ensureLocal(stmt.expr.toEtsEntity())
+                    etsStatements += EtsReturnStmt(
+                        location = stub,
+                        returnValue = returnValue,
+                    )
+                }
+
+                is BlockIf -> {
+                    val condition = ensureLocal(stmt.condition.toEtsEntity())
+                    etsStatements += EtsIfStmt(
+                        location = stub,
+                        condition = condition,
+                    )
+                }
+            }
+        }
+
+        if (etsStatements.isEmpty()) {
+            etsStatements += EtsNopStmt(location = stub)
+        }
+
+        return BasicBlock(
+            id = id,
+            statements = etsStatements,
+        )
+    }
+
+    private val stub
+        get() = EtsStmtLocation.stub(method)
+
+    private fun Expr.toEtsEntity(): EtsEntity = when (this) {
+        is Local -> EtsLocal(
+            name = name,
+            type = EtsUnknownType, // TODO
+        )
+
+        is Parameter -> EtsParameterRef(
+            index = index,
+            type = EtsUnknownType, // TODO
+        )
+
+        ThisRef -> EtsThis(
+            type = EtsUnknownType, // TODO
```

Comment:
Why todos?

Reply:
Because we probably can pass some outer class (enclosing the method for which for build the CFG) and use it as a type for ThisRef.

Reply:
Actually, this is just a DSL CFG, so probably we even do not need to solve its problems. DSL for programs lacks many typing information.

Reply:
Actually, this CFG builder is for converting from "DSL CFG" to "ETS CFG". DSL for programs lacks many typing information, and this is intentional and pretty much expected, so probably we do not need to try fixing all its problems.